src/aspire/basis/ffb_2d.py

```diff
@@ -160,9 +161,10 @@ def evaluate(self, v):
 
         # perform inverse non-uniformly FFT transform back to 2D coordinate basis
         freqs = m_reshape(self._precomp["freqs"], (2, n_r * n_theta))
-        x = np.zeros((self.sz[0], self.sz[1], n_data), dtype=v.dtype)
-        for isample in range(0, n_data):
-            x[..., isample] = 2*np.real(anufft3(pf[:, isample], 2 * pi * freqs, self.sz))
+
+        # TODO: Address axis swapping once C major in memory.
+        pfc = np.swapaxes(pf, 0, 1)
+        x = 2 * anufft3(pfc, 2 * pi * freqs, self.sz, real=True, many=n_data)
```

Comment:
To be consistent with what we did recently in PolarBasis2D class, probably we should explicitly set the data type of `x` as below: 
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/develop/src/aspire/basis/polar_2d.py#L83
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/develop/src/aspire/basis/polar_2d.py#L98

Reply:
`anufft3` should passthrough the precision corresponding to its input (in this case pfc). That is what we want everything in ASPIRE to do and why you were asked to fix the polar_2d code.  

If this is not happening, I want to know about it.  Do you have a case where this is not happening? What is it and what are the types of `x` and `pfc`?

Thanks.

Reply:
#148 ?

Reply:
Well there we go. Thought this discussion seemed familiar.

Reply:
I think in the base class @garrettwrong added the `dtype` attribute. That is the main reason we made changes in polar_2d.py and made the explicitly setting. It might be better to solve here since it is a small change. #148 is a general discussion instead of detailed solution to a particular class. But I am OK if you think you want come back later.

Reply:
False. The main reason I made any dtype changes anywhere is someone was really sloppy about types in their work with no intent to take responsibility for it, and it was breaking things. These included the stability of some unit tests, and also code that would fail without explicit extra copies (the hacks sprinkled around ASPIRE were not always _just_ for C/F ordering), and code that just didn't work when changing precision. Previously I touched things up just enough to get down the road, had discussions, and made issues. ASPIRE needs to be totally gone through top to bottom. We've discussed this many times. It's a problem. Between this, the ordering, on going debates about architecture... some shops would toss the whole project, start again from a whiteboard and designs... accept the losses and try to learn from the failures...

_Again_, `x` should be a precision corresponding to `pf`.  This passthrough is meant to be like numpy. That is `aspire.nufft`'s responsibility. FFB code is FFB's. If a `nufft` wrapper is not doing what it's supposed to, I want to know, but if you follow code below, I'm pretty doubtful. Instead of providing a meaningful and true counterexample as you were asked, you've just wasted my time with unrelated remarks _trying to deflect responsibility_.  The fact we can't get through simple feature add PRs without trying to scope creep into global technical debt in prior work is a systemic failure. It needs to stop. If you feel the dtypes should be fixed, take the issue, get with an experienced engineer to help you and fix your code. Otherwise, someone else will later, and this is not helpful to me, the PR, or the project.  I think things should get fixed too, but instead of doing something useful in an engineering sense, we're doing this. _It is totally sad._

The types input to `anufft` are checked by the `cufinufft` layer. Below is a line that checks the signal against the dtype inferred from other arguments in the "plan" (which are packed into `self`).  Reading this, it would seem to me the code would throw an exception if the numeric arrays were not a corresponding precision. There is a corresponding similar line for `transform`. You probably wouldn't know this, but the external cufinufft python code also has some similar checks, including unit tests that confirm the asserts are raised when fed the wrong type. We should be doing that stuff, but again, we're stuck in the mud all the time.

```
 assert signal.dtype == self.complex_dtype or signal.dtype == self.dtype
```

While I appreciate some review conversations for their insight, discussion, and improvements they can also be huge wastes of time when used for other purposes.

Reply:
As I said, let's leave this for now. We can't fix all of ASPIRE in a single PR. The point is to add cufinufft support, so let's focus on how it does that. @junchaoxia, if you feel that #148 doesn't address this particular problem, please make another issue so we can discuss it there and work on a solution.